pkg/handlers/v1/notification.go

```diff
@@ -0,0 +1,71 @@
+package v1
+
+import (
+	"context"
+
+	"github.com/asecurityteam/nexpose-scan-notifier/pkg/domain"
+	"github.com/asecurityteam/nexpose-scan-notifier/pkg/logs"
+)
+
+// Output contains a list of completed Nexpose scans.
+type Output struct {
+	Response []scanNotification `json:"response"`
+}
+
+// scanNotification represents
+type scanNotification struct {
+	ScanID string `json:"scanID"`
+	SiteID string `json:"siteID"`
+}
+
+// NotificationHandler takes a duration and returns a list of completed scans.
+type NotificationHandler struct {
+	ScanFetcher      domain.ScanFetcher
+	TimestampFetcher domain.TimestampFetcher
+	TimestampStorer  domain.TimestampStorer
+	Producer         domain.Producer
+	LogFn            domain.LogFn
+}
+
+// Handle queries for completed scans since the last invocation, produces all
+// completed scans to a queue, and returns the list of completed scans.
+func (h *NotificationHandler) Handle(ctx context.Context) (Output, error) {
+	logger := h.LogFn(ctx)
+
+	lastScanTimestamp, err := h.TimestampFetcher.FetchTimestamp(ctx)
+	switch err.(type) {
+	case nil:
+	case domain.TimestampNotFound:
+	default:
+		logger.Error(logs.StorageFailure{Reason: err.Error()})
+		return Output{}, err
+	}
+
+	scans, err := h.ScanFetcher.FetchScans(ctx, lastScanTimestamp)
+	if err != nil {
+		logger.Error(logs.ScanFetcherFailure{Reason: err.Error()})
+		return Output{}, err
+	}
+
+	scanNotifications := make([]scanNotification, len(scans))
+	for i, scan := range scans {
+		// Produce completed scans to a queue
+		err := h.Producer.Produce(ctx, scan)
+		if err != nil {
+			logger.Error(logs.ProducerFailure{Reason: err.Error()})
+			return Output{}, err
+		}
+		if err := h.TimestampStorer.StoreTimestamp(ctx, scan.Timestamp); err != nil {
+			return Output{}, err
+		}
```

Comment:
I like that you're storing the scan timestamp, not the API invocation time, for the following reasons:

1. Clock skew between the nexpose console machine and this service could cause gaps in scan completion events (though the probability of this occurring is admittedly low).
2. Because of the way Nexpose stores timestamps - as naive timestamps, based on whatever timezone is set on the machine running the scan - yes, this is incredibly unfortunate, but we must deal with it, the real time may not line up with whatever time Nexpose reports.
3. Using the scan timestamps allows you to move the marker for each event successfully produced, while not setting the marker and bailing out on any `Produce` failures. **However**, this approach only works in the partial case under one of the following conditions:
    i. The `FetchScans` method of the `ScanFetcher` interface returns the scans sorted in ascending order by timestamp. If you go this route, this should be a documented part of the interface.
    ii. The handler sorts the scans returned by the interface in the same manner, prior to producing them.

Also, you should update the `api.yaml` descriptions to note that you are fetching scans since the last known successfully processed scan timestamp.

Task: document the sorting order in the `ScanFetcher` interface, OR sort the scans in the handler.
Task: update the `api.yaml` to indicate the correct meaning of the timestamp.

Reply:
Good call. I think I prefer sorting the scans in the handler, not relying on the fetcher to return a sorted list.

Reply:
Fixed. Add an explicit sort in the handler and updated api.yaml